File: KKY_Tool_Revit_2019-2023/Infrastructure/ExcelStyleHelper.vb
Imports System
Imports System.Runtime.CompilerServices
Imports NPOI.SS.UserModel

Namespace Infrastructure

    Public Module ExcelStyleHelper

        Public Enum RowStatus
            None = 0
            Info = 1
            Warning = 2
            [Error] = 3
        End Enum

        Private Class StyleSet
            Public HeaderStyle As ICellStyle
            Public InfoStyle As ICellStyle
            Public WarningStyle As ICellStyle
            Public ErrorStyle As ICellStyle
        End Class

        Private ReadOnly _wbStyles As New ConditionalWeakTable(Of IWorkbook, StyleSet)()
        Private ReadOnly _wbNoWrapStyles As New ConditionalWeakTable(Of IWorkbook, StyleSet)()

        Private Function GetStyleSet(wb As IWorkbook) As StyleSet
            If wb Is Nothing Then Throw New ArgumentNullException(NameOf(wb))
            Return _wbStyles.GetValue(wb, Function(key) CreateStyleSet(key))
        End Function

        Private Function GetNoWrapStyleSet(wb As IWorkbook) As StyleSet
            If wb Is Nothing Then Throw New ArgumentNullException(NameOf(wb))
            Return _wbNoWrapStyles.GetValue(wb, Function(key) CreateStyleSet(key))
        End Function

        Private Function CreateStyleSet(wb As IWorkbook) As StyleSet
            Dim setx As New StyleSet()

            ' Header
            Dim header = wb.CreateCellStyle()
            Dim hf = wb.CreateFont()
            hf.IsBold = True
            hf.Color = IndexedColors.White.Index
            header.SetFont(hf)
            header.Alignment = HorizontalAlignment.Center
            header.VerticalAlignment = VerticalAlignment.Center
            header.WrapText = False
            header.FillForegroundColor = IndexedColors.Grey50Percent.Index
            header.FillPattern = FillPattern.SolidForeground
            header.BorderBottom = BorderStyle.Thin
            header.BorderTop = BorderStyle.Thin
            header.BorderLeft = BorderStyle.Thin
            header.BorderRight = BorderStyle.Thin
            setx.HeaderStyle = header

            ' Info
            setx.InfoStyle = CreateFillStyle(wb, IndexedColors.LightCornflowerBlue.Index)

            ' Warning
            setx.WarningStyle = CreateFillStyle(wb, IndexedColors.LightYellow.Index)

            ' Error
            setx.ErrorStyle = CreateFillStyle(wb, IndexedColors.Rose.Index)

            Return setx
        End Function

        Private Function CreateFillStyle(wb As IWorkbook, fillColor As Short) As ICellStyle
            Dim s = wb.CreateCellStyle()
            s.WrapText = False
            s.VerticalAlignment = VerticalAlignment.Center
            s.FillForegroundColor = fillColor
            s.FillPattern = FillPattern.SolidForeground
            s.BorderBottom = BorderStyle.Thin
            s.BorderTop = BorderStyle.Thin
            s.BorderLeft = BorderStyle.Thin
            s.BorderRight = BorderStyle.Thin
            Return s
        End Function

        Public Function GetHeaderStyle(wb As IWorkbook) As ICellStyle
            Return GetStyleSet(wb).HeaderStyle
        End Function

        Public Function GetHeaderStyleNoWrap(wb As IWorkbook) As ICellStyle
            Return GetNoWrapStyleSet(wb).HeaderStyle
        End Function

        Public Function GetRowStyle(wb As IWorkbook, status As RowStatus) As ICellStyle
            Dim setx = GetStyleSet(wb)
            Select Case status
                Case RowStatus.Info
                    Return setx.InfoStyle
                Case RowStatus.Warning
                    Return setx.WarningStyle
                Case RowStatus.Error
                    Return setx.ErrorStyle
                Case Else
                    Return Nothing
            End Select
        End Function

        Public Function GetRowStyleNoWrap(wb As IWorkbook, status As RowStatus) As ICellStyle
            Dim setx = GetNoWrapStyleSet(wb)
            Select Case status
                Case RowStatus.Info
                    Return setx.InfoStyle
                Case RowStatus.Warning
                    Return setx.WarningStyle
                Case RowStatus.Error
                    Return setx.ErrorStyle
                Case Else
                    Return Nothing
            End Select
        End Function

        Public Sub ApplyStyleToRow(row As IRow, colCount As Integer, style As ICellStyle)
            If row Is Nothing OrElse style Is Nothing Then Return

            Dim lastCol As Integer = Math.Max(0, colCount - 1)
            For c As Integer = 0 To lastCol
                Dim cell = row.GetCell(c)
                If cell Is Nothing Then
                    cell = row.CreateCell(c)
                    cell.SetCellValue("")
                End If
                cell.CellStyle = style
            Next
        End Sub

    End Module

End Namespace

File: KKY_Tool_Revit_2019-2023/Infrastructure/ExcelStyleHelper.cs
using System;
using System.Runtime.CompilerServices;
using NPOI.SS.UserModel;

namespace KKY_Tool_Revit.Infrastructure
{
    public static class ExcelStyleHelper
    {
        public enum RowStatus
        {
            None = 0,
            Info = 1,
            Warning = 2,
            Error = 3
        }

        private sealed class StyleSet
        {
            public ICellStyle HeaderStyle;
            public ICellStyle InfoStyle;
            public ICellStyle WarningStyle;
            public ICellStyle ErrorStyle;
        }

        private static readonly ConditionalWeakTable<IWorkbook, StyleSet> WbStyles = new ConditionalWeakTable<IWorkbook, StyleSet>();
        private static readonly ConditionalWeakTable<IWorkbook, StyleSet> WbNoWrapStyles = new ConditionalWeakTable<IWorkbook, StyleSet>();

        private static StyleSet GetStyleSet(IWorkbook wb)
        {
            if (wb == null) throw new ArgumentNullException(nameof(wb));
            return WbStyles.GetValue(wb, CreateStyleSet);
        }

        private static StyleSet GetNoWrapStyleSet(IWorkbook wb)
        {
            if (wb == null) throw new ArgumentNullException(nameof(wb));
            return WbNoWrapStyles.GetValue(wb, CreateStyleSet);
        }

        private static StyleSet CreateStyleSet(IWorkbook wb)
        {
            var setx = new StyleSet();

            var header = wb.CreateCellStyle();
            var hf = wb.CreateFont();
            hf.IsBold = true;
            hf.Color = IndexedColors.White.Index;
            header.SetFont(hf);
            header.Alignment = HorizontalAlignment.Center;
            header.VerticalAlignment = VerticalAlignment.Center;
            header.WrapText = false;
            header.FillForegroundColor = IndexedColors.Grey50Percent.Index;
            header.FillPattern = FillPattern.SolidForeground;
            header.BorderBottom = BorderStyle.Thin;
            header.BorderTop = BorderStyle.Thin;
            header.BorderLeft = BorderStyle.Thin;
            header.BorderRight = BorderStyle.Thin;
            setx.HeaderStyle = header;

            setx.InfoStyle = CreateFillStyle(wb, IndexedColors.LightCornflowerBlue.Index);
            setx.WarningStyle = CreateFillStyle(wb, IndexedColors.LightYellow.Index);
            setx.ErrorStyle = CreateFillStyle(wb, IndexedColors.Rose.Index);

            return setx;
        }

        private static ICellStyle CreateFillStyle(IWorkbook wb, short fillColor)
        {
            var s = wb.CreateCellStyle();
            s.WrapText = false;
            s.VerticalAlignment = VerticalAlignment.Center;
            s.FillForegroundColor = fillColor;
            s.FillPattern = FillPattern.SolidForeground;
            s.BorderBottom = BorderStyle.Thin;
            s.BorderTop = BorderStyle.Thin;
            s.BorderLeft = BorderStyle.Thin;
            s.BorderRight = BorderStyle.Thin;
            return s;
        }

        public static ICellStyle GetHeaderStyle(IWorkbook wb) => GetStyleSet(wb).HeaderStyle;

        public static ICellStyle GetHeaderStyleNoWrap(IWorkbook wb) => GetNoWrapStyleSet(wb).HeaderStyle;

        public static ICellStyle GetRowStyle(IWorkbook wb, RowStatus status)
        {
            var setx = GetStyleSet(wb);
            switch (status)
            {
                case RowStatus.Info: return setx.InfoStyle;
                case RowStatus.Warning: return setx.WarningStyle;
                case RowStatus.Error: return setx.ErrorStyle;
                default: return null;
            }
        }

        public static ICellStyle GetRowStyleNoWrap(IWorkbook wb, RowStatus status)
        {
            var setx = GetNoWrapStyleSet(wb);
            switch (status)
            {
                case RowStatus.Info: return setx.InfoStyle;
                case RowStatus.Warning: return setx.WarningStyle;
                case RowStatus.Error: return setx.ErrorStyle;
                default: return null;
            }
        }

        public static void ApplyStyleToRow(IRow row, int colCount, ICellStyle style)
        {
            if (row == null || style == null) return;

            var lastCol = Math.Max(0, colCount - 1);
            for (var c = 0; c <= lastCol; c++)
            {
                var cell = row.GetCell(c) ?? row.CreateCell(c);
                if (cell.CellType == CellType.Blank)
                {
                    cell.SetCellValue(string.Empty);
                }
                cell.CellStyle = style;
            }
        }
    }
}
